### PR TITLE
veOCEAN Delegation

### LINKED ISSUE #303: Implement delegation
There are other designs we've looked at (CVX), let's design a component and implement it.

### DoD:
- [x] Design/Dev delegation component
- [x] User is able to delegate veOCEAN and get info on it
- [x] Do full E2E test (FE work). Blocked by "E2E test on goerli" issue

### PR DESCRIPTION
**Pull request by KatunaNorbert**
Fixes #303.

### CODE REVIEW

**Comment by idiom-bytes** [comment]
Tested, worked great!
- Can you please write a markdown documentation for this feature?
- Assume that this will be part of a larger doc called veOCEAN [preview here](https://docs.google.com/document/d/1Y-1Mmmu49VZ7zZ7RnsCDfwj22dQK15lmyQFTG0QE2KA/edit)

**Review by idiom-bytes — APPROVED** [review]
LGTM

**Review by jamiehewitt15 — APPROVED** [review]
Looks good, seems to work well

**Comment by EnzoVezzaro** [comment]
I found few issue, sorry for getting into the weeds 😅

1) For more clarity, we can add a little to this tooltip message: "The date on which your delegation will expiry." for me is vague, we can just add "The date on which your delegation will expiry (lock end date)." or improve a little the messaging around this.

Screenshot 2023-03-15 at 08 11 29

Also, on the Delegated account I don't see the date:

Screenshot 2023-03-15 at 08 23 07

2) I tried to delegate from 2 accounts to just one delegation wallet. 

The first tx seems fine: the delegation wallet got its delegation power, and the account delegating had his removed. All good

The delegation wallet can't delegate to another account (I though they could) and we see this message: "You can cancel your delegation at any time.", but we don't have a way to cancel as delegated account. 

Screenshot 2023-03-15 at 08 32 55

Then, from a third account I delegated again to the delegation wallet:

Screenshot 2023-03-15 at 08 25 54

All the allocation power seems to be sum up fine, but I have a lot questions with it:
- what happen if the delegated account has allocated to some assets, and then their allocation power gets cancelled?
- if I receive some allocation power means that I can't allocate to others? 
- If I receive some allocation power when I allocate, I'm allocating my allocation power? the allocation power I received? If what I receive got cancelled, how do we know how's allocation power we need to cancel? For example: a delegated account has receive allocation power from 4-5 different wallets. 

3) when you cancel a delegation, your stats goes to 0, but if you refresh the numbers are fine.

Screenshot 2023-03-15 at 08 30 28

This is what I have right now, I'll keep looking later today.

**Comment by EnzoVezzaro** [comment]
UPDATES:

> * what happen if the delegated account has allocated to some assets, and then their allocation power gets cancelled?

the allocation power is proportional by the % allocated. If some allocation power gets removed, the total allocation power get updated, but the % allocated into the assets stays.

> * if I receive some allocation power means that I can't allocate to others?

Yes, I should be able to. A delegated account can delegate their OWN allocation power to others (they can't delegate what they had receive).

> * If I receive some allocation power, when I allocate I'm allocating my own allocation power also? or just the allocation power I received? If what I receive gets cancelled, how do we know how's allocation power we need to cancel? For example: a delegated account has receive allocation power from 4-5 different wallets.

the allocation power is proportional by the % allocated. If some allocation power gets removed, the total allocation power get updated, but the % allocated into the assets stays.

**Comment by KatunaNorbert** [comment]
@EnzoVezzaro thanks for looking deeper into this and asking the questions.
Beside the tooltip improvements, can you list what other issues you discovered and questions that are left if there is any?

**Comment by KatunaNorbert** [comment]
What is veOCEAN Delegation
-Delegation allows you to give your veOCEAN allocation power to another wallet address for a limited period of time.

Why to delegate
-Higher APY by delegation to wallet addresses that manages your allocation in the most efficient way.
-No need of reallocating and pay transaction fees related to that.
-If are locked with multiple wallets, you can use just one of them to manage your allocations.

How delegation works
-When delegation you will give away for a limited period of time 100% of your veOCEAN allocation power. After that you wouldn't be able to manage your allocations anymore, until delegation expires.
-Your delegation expiration date is the veOCEAN lock end date of the moment when the delegation is executed. You might want to extend your lock before delegating.
-You can cancel your delegation at any time.

How do I get rewards
-After you are delegated, rewards are going to go to the wallet address that you delegated to.
-The delegation receiver is going to take a cut of the rewards and send the rest of the rewards to you.

How to delegate
-Go to [DF Portal](https://df.oceandao.org/delegate) and select a wallet address that you want to delegate to.
-Click the Delegate button and sign the transaction. After this you can see information about your delegation in the My Delegations component.
-Cancel delegation if you want to get your allocation power back before the delegation expiry date.

How to manage received delegations
-If you received veOCEAN allocation power from other wallet addresses you will receive their rewards and will be in charge of sending rewards to delegators
-You can’t delegate the veOCEAN that you received from delegates, only veOCEAN that you received from your lock.

**Comment by EnzoVezzaro** [comment]
> can you list what other issues you discovered and questions that are left if there is any?

I though I had answer this 😅 so, basically there's still this couple of things:

> Also, on the Delegated account I don't see the date (after receiving allocation from another account):

and this:

> when you cancel a delegation, your stats goes to 0, but if you refresh the numbers are fine.

**Comment by jamiehewitt15** [comment]
One thing that's unclear to me, is how delegation affects the number of active rewards that you will receive. The current page seems to imply that it will increase your APY as you avoid gas fees, but presumably, the person that you delegate to takes some sort of fee to cover gas and provide them with an incentive to do this?

**Comment by KatunaNorbert** [comment]
@jamiehewitt15 If you delegate to a wallet that efficiently allocates you can get a much higher APY than your current one. If you have your own method that gives you maximum APY than it's not going to help because of the reward cut.

**Comment by KatunaNorbert** [comment]
I don't think we can, it depends on the part you delegated to

**Comment by jamiehewitt15** [comment]
> I don't think we can, it depends on the part you delegated to

hmm ok, is there a maximum they can set?

It might still be useful to say somewhere "Please note that the account you delegate to will take a percentage of the rewards"

**Comment by KatunaNorbert** [comment]
I guess we could build something on the df-py where we have a constant percentage for receiver that does the calculation and split the rewards. This was the delegation receiver doesn't have that much control and work to do. But then If I have multiple accounts and delegate to a main one, I might just want to have all the rewards in 1 place and sign just 1 transaction to claim them.

**Comment by KatunaNorbert** [comment]
@EnzoVezzaro pushed some changes, hope this fixes your issue

**Review by jamiehewitt15 — DISMISSED** [review]
I've tested a couple of times with different accounts and it all seems to work well. The messaging is also better now.

**Comment by KatunaNorbert** [comment]
@EnzoVezzaro Good feedback!
1. I will push a fix soon
2. I don't think we are going to handle this on the first version. You might get multiple delegations so in order to show when the received delegation it's going to end we would have to show all users who delegated to you and the end date for each one. So for now we only display your delegation end that, because you can only delegate to 1 person.

**Review by jamiehewitt15 — APPROVED** [review]

**Review by EnzoVezzaro — DISMISSED** [review]
LGTM

**Review by EnzoVezzaro — DISMISSED** [review]

**Comment by idiom-bytes** [comment]
1. @trizin I agree. @KatunaNorbert can you adjust and deploy?
1. One ask I have. Can we move "Delegate" to the right of Rewards?
1. Please make @nickscavuzzo33 aware that we're deploying Delegation mechanisms.

File: src/components/delegation/Delegate.svelte [hunk 1]
```diff
@@ -119,6 +119,10 @@
       You will delegate your entire veOCEAN allocation until your current lock
       will end, and you can cancel at any time.
     </p>
+    <p class="message">
+      The wallet address you delegate to it's going to receive 100% of your
```
Comment (jamiehewitt15) [inline src/components/delegation/Delegate.svelte:123]:
```suggestion
      The wallet address you delegate to is going to receive 100% of your
```

**Review by EnzoVezzaro — DISMISSED** [review]
found just a small typo. All good, great job 👍🏽

**Review by EnzoVezzaro — DISMISSED** [review]

**Review by jamiehewitt15 — APPROVED** [review]
Looks ready to go to me

**Comment by idiom-bytes** [comment]
Dependency on subpgrah release:
https://github.com/oceanprotocol/ocean-subgraph/pull/623

**Comment by jamiehewitt15** [comment]
Looks like this has conflicts on it now. Can you resolve them @KatunaNorbert?

**Review by jamiehewitt15 — APPROVED** [review]
I've tested this several times and it's very reliable, I don't see any issues. The front end is ready to go, let's just double-check all of the other components before merging

File: src/components/delegation/Delegate.svelte [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        : delegateVeOcean(values);
+    },
+  });
+</script>
+
+<div class={`container`}>
+  <Card title="Delegate">
+    <p class="message">
+      You will delegate your entire veOCEAN allocation until your current lock
+      will end, and you can cancel at any time.
```
Comment (idiom-bytes) [inline src/components/delegation/Delegate.svelte:122]: "You will delegate your entire veOCEAN allocation until your current lock ends. You can cancel at any time."

File: src/components/delegation/Delegate.svelte [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    </p>
+    <p class="message">
+      The wallet address you delegate to is going to receive 100% of your
+      rewards.
```
Comment (idiom-bytes) [inline src/components/delegation/Delegate.svelte:126]: "The wallet address you delegate to is going to receive 100% of your Active Rewards."

File: src/components/delegation/DelegationPortal.svelte [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        ...activeDelegation,
+        createId: createId,
+      };
+    });
+  }
+</script>
+
+<div class={`container`}>
+  <h2 class="title">
+    Allow other wallet to manage your veOCEAN allocation by delegating.
```
Comment (idiom-bytes) [inline src/components/delegation/DelegationPortal.svelte:52]: "Allow another wallet to manage your veOCEAN allocation by delegating."

File: src/components/delegation/DelegationPortal.svelte [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  </h2>
+  <p class="message">
+    Maximize your active APY and skip transaction fees by delegating to a wallet
+    that efficiently manages your allocation power.
```
Comment (idiom-bytes) [inline src/components/delegation/DelegationPortal.svelte:56]: "Maximize your **Active** APY and skip transaction fees by delegating to a wallet that efficiently manages your allocation power."

File: src/components/delegation/Delegate.svelte (hunk 3 above)
Comment (idiom-bytes) [reply src/components/delegation/Delegate.svelte:126]: Expand text: "Please update your delegation whenever you obtain more veOCEAN."

**Review by idiom-bytes — DISMISSED** [review]
Looks great!

File: src/components/delegation/Delegate.svelte (hunk 2 above)
Comment (KatunaNorbert) [reply src/components/delegation/Delegate.svelte:122]: done

File: src/components/delegation/Delegate.svelte (hunk 3 above)
Comment (KatunaNorbert) [reply src/components/delegation/Delegate.svelte:126]: done

File: src/components/delegation/DelegationPortal.svelte (hunk 4 above)
Comment (KatunaNorbert) [reply src/components/delegation/DelegationPortal.svelte:52]: done

File: src/components/delegation/DelegationPortal.svelte (hunk 5 above)
Comment (KatunaNorbert) [reply src/components/delegation/DelegationPortal.svelte:56]: done

**Comment by idiom-bytes** [comment]
Looks great, thanks for the updates to the copy.

**Review by idiom-bytes — DISMISSED** [review]
Reviewed, looks good!

**Pull request merged**